keras_retinanet/preprocessing/csv_generator.py

```diff
@@ -0,0 +1,117 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+
+import keras_retinanet
+
+import cv2
+import os
+import numpy as np
+from PIL import Image
+
+import cv2
+import csv
+
+
+class CSVGenerator(keras_retinanet.preprocessing.Generator):
+    def __init__(
+        self,
+        csv_file,
+        set_name,
+        *args,
+        **kwargs
+    ):
+        self.csv_file             = csv_file
+        self.set_name             = set_name
+
+        self.classes              = {}
+        self.image_names          = []
+        self.image_data           = {}
+
+        # csv with img_filepath, class, x1, y1, x2, y2, dataset
+        with open(csv_file, 'rb') as f_in:
+            csvreader = csv.reader(f_in, delimiter=',')
+            for row in csvreader:
+
+                img_filepath, x1, y1, x2, y2, classname, dataset = row
+
+                # populate the classes from the CSV automatically and assign them an ID
+                if classname not in self.classes:
+                    self.classes[classname] = len(self.classes)
+
+                if dataset != set_name:
+                    continue
+
+                if img_filepath not in self.image_names:
+                    self.image_names.append(img_filepath)
+                    self.image_data[img_filepath] = []
+                self.image_data[img_filepath].append({'x1':int(x1), 'x2':int(x2), 'y1': int(y1), 'y2':int(y2), 'class':classname})
+
+        self.labels = {}
+        for key, value in self.classes.items():
+            self.labels[value] = key
+
+        super(CSVGenerator, self).__init__(*args, **kwargs)
+
+    def size(self):
+        return len(self.image_names)
+
+    def num_classes(self):
+        return len(self.classes)
+
+    def name_to_label(self, name):
+        return self.classes[name]
+
+    def label_to_name(self, label):
+        return self.labels[label]
+
+    def image_aspect_ratio(self, image_index):
+        path  = self.image_names[image_index]
+        # PIL is fast for metadata
+        image = Image.open(path)
+        return float(image.width) / float(image.height)
+
+    def load_image(self, image_index):
+        path = self.image_names[image_index]
+        return cv2.imread(path)
+
+    def load_annotations(self, image_index):
+        boxes = np.zeros((0, 5))
+
+        path  = self.image_names[image_index]
+        image = Image.open(path)
+
+        width = float(image.width)
+        height = float(image.height)
+
+        annots = self.image_data[path]
+
+        for annot in annots:
+
+            box = np.zeros((1, 5))
+
+            class_name = annot['class']
+
+            box[0, 4] = self.name_to_label(class_name)
+
+            bndbox = [annot['x1'], annot['y1'], annot['x2'], annot['y2']]
+            box[0, 0] = float(bndbox[0]) - 1
+            box[0, 1] = float(bndbox[1]) - 1
+            box[0, 2] = float(bndbox[2]) - 1
+            box[0, 3] = float(bndbox[3]) - 1
```

Comment:
@yhenon Does this mean that you expect indexing of pixels in csv file to be 1 based (starting from 1) and here you transform it to 0 based, i.e. the first pixel has index 0?

Reply:
this matches the behaviour in keras_retinanet/preprocessing/pascal_voc.py

I guess we can either: (a) define the csv format to index starting at 0 and remove the `-1` here, or (b) define the csv format to index starting at 1 and keep the `-1` here.

Reply:
Personally I feel we shouldn't copy weird choices made by pascal VOC. One-based indexing is odd. I'm in favor of dropping the -1 and defining the CSV format using zero based indices.

Reply:
As long as it's documented somewhere, e.g. in readme, both options are good for me.

Reply:
fixed as per @de-vri-es comment which I agree with